### ISSUE TITLE
The value of a parameter is not passed to submodel (regression)

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#9825]
### Description
I have a model in which the value of a parameter is not passed to a submodel: it stays to zero. This model works well with Dymola, and OM up to 1.18.0. 
It shows the issue with 1.20.0.beta4 and 1.19.2.

### Steps to Reproduce
1. Load Test.mo:
[Test.mo.txt](https://github.com/OpenModelica/OpenModelica/files/10130875/Test.mo.txt)
2. run M and look at the result. As you can see in the first screenshot below, the value of wMax, which in drive is 1200, inside tauLim stays to zero. In the second screenshot below you see the expected behaviour.

### Expected Behavior
wMax inside limTau should be 1200 rad/s (second screenshot below)

### Screenshots
With 1.20.0:
![screen](https://user-images.githubusercontent.com/7373030/205010592-8205dc11-e550-4165-9bcf-c235ad134033.png)
With 1.18:
![1 18](https://user-images.githubusercontent.com/7373030/205012105-3961e81a-21be-4b73-b4fa-41bb739ad33b.png)

**Comment by casella** [comment]
@ceraolo, one side comment before getting to the real issue: the attached package has a uses(Modelica) annotation in _the model_ and not _in the package itself_, which doesn't really work. You can't have different models with different dependencies in the same library, that would be impossible to manage. 

Did you build this test package with OMEdit?

**Comment by ceraolo** [comment]
I added that uses by hand.
Ok, that uses annotation should have been in the package.
Nevertheless, this use of _uses_ is well accepted by OM (at least since 1.18.0) and Dymola. So, it is a bit weird but considered acceptable. I did not check MS.

**Comment by casella** [comment]
@ceraolo thanks for reporting, this is really a weird bug. I checked with the declarative debugger, and the parameter seems to be properly propagated:
```
170  parameter  (assign) drive.limTau.wMax := drive.wMax
```
The M_08_bnd.c file contains:
```C
/*
equation index: 170
type: SIMPLE_ASSIGN
drive.limTau.wMax = drive.wMax
*/
OMC_DISABLE_OPT
static void M_eqFunction_170(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  const int equationIndexes[2] = {1,170};
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = (data->simulationInfo->realParameter[68]/* drive.wMax PARAM */) ;
  TRACE_POP
}
```
and this function is then called by the `M_updateBoundParameters_0()` function.

The parameter `drive.toElePow.wMax` has an identical treatment in `M_eqFunction_151()` 
```C
/*
equation index: 151
type: SIMPLE_ASSIGN
drive.toElePow.wMax = drive.wMax
*/
OMC_DISABLE_OPT
static void M_eqFunction_151(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  const int equationIndexes[2] = {1,151};
  (data->simulationInfo->realParameter[65]/* drive.toElePow.wMax PARAM */)  = (data->simulationInfo->realParameter[68]/* drive.wMax PARAM */) ;
  TRACE_POP
}
```

but is handled correctly during simulation, i.e. it gets a value of 1200 rad/s.

I also checked the M_init.xml file, where the two parameters are treated identically.

```xml
  <ScalarVariable
    name = "drive.limTau.wMax"
    valueReference = "1072"
    description = "Maximum speed"
    variability = "parameter" isDiscrete = "true"
    causality = "calculatedParameter" isValueChangeable = "false"
    alias = "noAlias"
    classIndex = "35" classType = "rPar"
    isProtected = "false" hideResult = ""
    fileName = "D:/Temp/OMEdit/Test.mo" startLine = "351" startColumn = "11" endLine = "351" endColumn = "103" fileWritable = "true">
    <Real fixed="true" useNominal="false" unit="rad/s" />
  </ScalarVariable>

  <ScalarVariable
    name = "drive.toElePow.wMax"
    valueReference = "1102"
    description = "Maximum machine speed"
    variability = "parameter" isDiscrete = "true"
    causality = "calculatedParameter" isValueChangeable = "false"
    alias = "noAlias"
    classIndex = "65" classType = "rPar"
    isProtected = "false" hideResult = ""
    fileName = "D:/Temp/OMEdit/Test.mo" startLine = "75" startColumn = "11" endLine = "75" endColumn = "96" fileWritable = "true">
    <Real fixed="true" useNominal="false" unit="rad/s" />
  </ScalarVariable>
```

@mahge, this is really, really weird. Could you please try to run the simulation executable with the debugger and figure out why `drive.limTau.wMax` is eventually evaluated to zero in simulation, while `drive.toElePow.wMax` is evaluated to 1200?

This is quite nasty bug, it should be fixed with high priority.

**Comment by casella** [comment]
I checked, the bug is present with the latest nightly, both under Windows 10 and under Ubuntu focal.

**Comment by perost** [comment]
> I added that uses by hand. Ok, that uses annotation should have been in the package. Nevertheless, this use of _uses_ is well accepted by OM (at least since 1.18.0) and Dymola. So, it is a bit weird but considered acceptable. I did not check MS.

`uses` that's not on a top-level package will be ignored by OM, but libraries that are used by the model will be loaded automatically regardless. In that case it just loads the latest installed version though. That's of course not the issue here, but just a FYI on how OM loads dependencies.

**Comment by mahge** [comment]
> @mahge, this is really, really weird. Could you please try to run the simulation executable with the debugger and figure out why drive.limTau.wMax is eventually evaluated to zero in simulation, while drive.toElePow.wMax is evaluated to 1200?

I am on it.

**Comment by max-privato** [comment]
Using min and max on parameters is useful to receive assert violation warnings at t=0, which can point errors on parameters' numerical values.

It is, however, a not so important feature, since it can be substituted by explicit asserts in as initial equations.
There are cases in which OM 1.20.0-beta4 works very well in this respect: e.g. in TextMax.mo:  [TestMax.mo.txt](https://github.com/OpenModelica/OpenModelica/files/10133269/TestMax.mo.txt)
where in simuMoldel we have literal max, and in simuModel2 max as a function of other parameters.

In my personal case, just removing min=powMax/tauMax value is very acceptable.
So, thank you for pointing this out.

Obviously, OM's zeroing of a parameter value in cases like this is a serious weakness, because it lets users confused (in the beginning) and disappointed (later).

**Comment by mahge** [comment]
> Using min and max on parameters is useful to receive assert violation warnings at t=0, which can point errors on parameters' numerical values.
> 
> It is, however, a not so important feature, since it can be substituted by explicit asserts in as initial equations.
> 
> There are cases in which OM 1.20.0-beta4 works very well in this respect: e.g. in TextMax.mo: [TestMax.mo.txt](https://github.com/OpenModelica/OpenModelica/files/10133269/TestMax.mo.txt) where in simuModel we have literal max, and in simuModel2 max as a function of other parameters.

In this case maybe we should investigate how it could be done in a way that does not persistently overwrite the values. 

I tried removing the resetting of the parameter to the start value (only update the min value) but it caused some failures in other models. Then I also (blindly) tried reversing the order in which `updateBoundParameters` and `updateBoundVariableAttributes` are called. That also caused some failures. Maybe it is a good idea to take a look at this with @kabdelhak, @AnHeuermann or @casella.

**Comment by casella** [comment]
> I added that uses by hand.

OK. I was worried that OMEdit added them, which shouldn't be the case.

> Ok, that uses annotation should have been in the package. Nevertheless, this use of _uses_ is well accepted by OM (at least since 1.18.0) and Dymola. So, it is a bit weird but considered acceptable. I did not check MS.

@ceraolo we try to stick as much as we can to the specification. Following _de facto_ non-standard features (of any tool, including OMC) is not a good policy for a Modelica tool: either we standardize them, or we don't support them. Otherwise we'd end up reverse-engineering stuff all the time, instead of implementing an open standard, which is already hard enough 😅 

The specification, Section 18.8.2, is very clear:
> uses(IDENT (version = VERSION-NUMBER [, versionBuild=INTEGER] [, dateModified=STRING] ) )
> Defines that classes within this **top-level class** uses version VERSION-NUMBER of classes within the top-level class IDENT. 

uses annotations are only considered within top level classes.

**Comment by casella** [comment]
> On the third line of the above output, you can see `drive.limTau.wMax` is being set to its min value. Updating the min value is **immediately followed** by setting it to its start value.
> 
> This is done in the function `Test_M_updateBoundVariableAttributes` which is evaluated **after** `Test_M_updateBoundParameters` (which calls `M_eqFunction_170` as @casella mentioned to set it `= drive.wMax = 1200` ).

Yes, I guess it's these lines of code:
```C
  infoStreamPrint(LOG_INIT, 1, "updating min-values");
  data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.min = (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */) ;
  infoStreamPrint(LOG_INIT_V, 0, "%s(min=%g)", data->modelData->realParameterData[35].info /* drive.limTau.wMax */.name, (modelica_real) data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.min);
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.start;
```
> The reason this is happening only for `drive.limTau.wMax` (and not e.g., for `drive.toElePow.wMax`) is because `wMax` in `block LimTau` (line: 351) is **given a min value** while the others are not.

OK.

> The reason this was not happening in 1.18 was because we recently changed things to traverse parameters and look for min, max ... value equations (Added in #8030, improved in #9612). My assumption in [#9213 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9213#issuecomment-1266857699) is probably wrong and this behavior of creating min, max ... value equation is **needed only for Dynamic Optimization** purposes. In which case we should only enable it when asked.

I don't think so. Min/max are useful for other things besides optimal control, e.g., optimizations of inStream() operators, or optimization of expressions: if you have x := y/z and z.max > 0, you can avoid DIVISION_MACRO in the generated code. 

Having more min/max information should never do harm. It's a good explanation of why we got this regression after your commits, but it doesn't mean we should revert them.

The real question is: why on Earth, after the code to set `drive.limTau.wMax.min = 250`, we also generate code that sets the parameter to its start attribute? This makes no sense to me, and in fact produces plain wrong results. BTW, this could be happening in other models and libraries as well.

**Comment by max-privato** [comment]
> @ceraolo we try to stick as much as we can to the specification. Following _de facto_ non-standard features (of any tool, including OMC) is not a good policy for a Modelica tool: either we standardize them, or we don't support them. Otherwise we'd end up reverse-engineering stuff all the time, instead of implementing an open standard, which is already hard enough 😅

I'm TOTALLY in favour of this approach. The contribution of OM to keep the Modelica standard stable is fundamental.
_Sic stantibus rebus_ I would even gladly accept OM not accepting at all the uses() annotation in models belonging to packages!

**Comment by ceraolo** [comment]
> @ceraolo we try to stick as much as we can to the specification. Following _de facto_ non-standard features (of any tool, including OMC) is not a good policy for a Modelica tool: either we standardize them, or we don't support them. Otherwise we'd end up reverse-engineering stuff all the time, instead of implementing an open standard, which is already hard enough 😅

I'm TOTALLY  in favour of this approach! The contribution of OM to keep modelica stable is fundamental.
_Sic stantibus rebus_, I would even gladly accept OM rejecting any uses() annotation outside top-level classes!

But I understand that, since up to now the policy was as in https://github.com/OpenModelica/OpenModelica/issues/9825#issuecomment-1333614232, it may not deemed advisable making changes to this that would break existing libraries

**Comment by ceraolo** [comment]
> @ceraolo we try to stick as much as we can to the specification. Following _de facto_ non-standard features (of any tool, including OMC) is not a good policy for a Modelica tool: either we standardize them, or we don't support them. Otherwise we'd end up reverse-engineering stuff all the time, instead of implementing an open standard, which is already hard enough 😅

I'm TOTALLY  in favour of this approach! The contribution of OM to keep modelica stable is fundamental.

However, if we read https://github.com/OpenModelica/OpenModelica/issues/9825#issuecomment-1333614232, in conjunction with specification:
>uses(IDENT (version = VERSION-NUMBER [, versionBuild=INTEGER] [, dateModified=STRING] ) )
Defines that classes within this top-level class uses version VERSION-NUMBER of classes within the top-level class IDENT.

it seems that there is coherency in the OM behaviour (and therefore in code having uses() outside top-level classes): in fact, https://github.com/OpenModelica/OpenModelica/issues/9825#issuecomment-1333614232 says that OM does not consider that uses() _to infer any version number_, just to load the library.
Otherwise one should consider even rejecting any uses() used outside top classes.
@casella what do you think?

**Comment by mahge** [comment]
> The real question is: why on Earth, after the code to set drive.limTau.wMax.min = 250, we also generate code that sets the parameter to its start attribute? This makes no sense to me, and in fact produces plain wrong results. BTW, this could be happening in other models and libraries as well.

This was added by 4f66de934b7d6ecfbde869651d3ee7853dd34c13. I think this was added with only variables in mind. Perhaps we need to revise what the approach should be for parameters. Even if we remove the resetting-to-start-value, the parameter will still have its min as value instead of what is assigned to it in the model. That is not what we want right?

**Comment by perost** [comment]
> it seems that there is coherency in the OM behaviour (and therefore in code having uses() outside top-level classes): in fact, [#9825 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9825#issuecomment-1333614232) says that OM does not consider that uses() _to infer any version number_, just to load the library. Otherwise one should consider even rejecting any uses() used outside top classes. @casella what do you think?

No, OM completely ignores `uses` on non top-level classes. But it loads libraries based on usage, i.e. for:
```Modelica
model M
  Modelica.Units.SI.Length l;
end M;
```
it will load the default version of Modelica when it encounters the `Modelica` identifier if it's not already loaded (as defined in [13.3](https://specification.modelica.org/master/packages.html#the-modelica-library-path-modelicapath)).

**Comment by casella** [comment]
> I'm TOTALLY in favour of this approach! The contribution of OM to keep modelica stable is fundamental.

👌 

> it seems that there is coherency in the OM behaviour (and therefore in code having uses() outside top-level classes): in fact, [#9825 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9825#issuecomment-1333614232) says that OM does not consider that uses() _to infer any version number_, just to load the library.

Not really. What @perost meant is that whenever OMC encounters a class that is not loaded, it will try to load it from MODELICAPATH, _regardless of uses annotations_, as specified by [Section 13.3](https://specification.modelica.org/maint/3.5/packages.html#the-modelica-library-path-modelicapath). So, it will load the most recent version found in the MODELICAPATH, or fail if it doesn't find it. Note that the most recent version could not be backward compatible with the actually needed one, so this "poor man's library management" is not very sophisticated.

 If the uses annotation is instead given _for the top level class_, then OMC uses it for a lot of extra things, e.g.:
- make sure the right version of the library is loaded
- actually load a more recent version _but only if it is backwards compatible_, based on information put in the conversion annotations
- make sure that all the right versions of the libraries needed by a library X are installed when library X is installed, and updated when the library X is updated - this is what the package manager does

> Otherwise one should consider even rejecting any uses() used outside top classes.

That's what we do already, we just ignore them, according to what the Specification says in Section 18.8.

**Comment by ceraolo** [comment]
> > it seems that there is coherency in the OM behaviour (and therefore in code having uses() outside top-level classes): in fact, [#9825 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9825#issuecomment-1333614232) says that OM does not consider that uses() _to infer any version number_, just to load the library. Otherwise one should consider even rejecting any uses() used outside top classes. @casella what do you think?
> 
> No, OM completely ignores `uses` on non top-level classes. But it loads libraries based on usage, i.e. for:
> 
> ```modelica
> model M
>   Modelica.Units.SI.Length l;
> end M;
> ```
> AHA.
I did not get this from your previous comment.

**Comment by casella** [comment]
> > The real question is: why on Earth, after the code to set drive.limTau.wMax.min = 250, we also generate code that sets the parameter to its start attribute? This makes no sense to me, and in fact produces plain wrong results. BTW, this could be happening in other models and libraries as well.
> 
> This was added by [4f66de9](https://github.com/OpenModelica/OpenModelica/commit/4f66de934b7d6ecfbde869651d3ee7853dd34c13). 

I had a look at 4f66de9, but I'm struggling to understand what's the point of it. Let's look at the [overall initialization()](https://github.com/OpenModelica/OpenModelica/blob/4f66de934b7d6ecfbde869651d3ee7853dd34c13/SimulationRuntime/c/simulation/solver/initialization/initialization.c#L547) function, assuming OMC_MINIMAL_RUNTIME is not defined. This is how I understand it:

- `setAllParamsToStart(data);` is called, which I guess sets all parameters to their start values. Then, the bound ones with default `fixed = true` will be overwritten, while the ones with `fixed = false` will retain their start values as initial guesses for the Newton solver during the solution of the initial equations
- if start values need to be imported then first the runtime sets bound parameters and attributes to the values hardcoded in the C source (and coming from the Modelica source code), then it imports then from a file, which overwrites them; if the file is incomplete, the missing values retain the values hardcoded in the Modelica and C code
- `setAllVarsToStart(data);` is called, which I guess sets all variables (not parameters) to their start values
- if there was no file to be imported, the runtime sets bound parameters and attributes to the values hardcoded in the C source and coming from the C code

Before 4f66de9, `setAllVarsToStart(data);` was finally called again, I'm not sure why. Probably because some start values may have been overwritten, but how? I have no idea. After that commit, this call was removed, but code was added to the template so that every time a bound variable (which apparently also includes bound _parameters_ such as `drive.limTau.wMax`) is updated, the variable is initialized to its start value. Given the title of 4f66de9, the idea was to "Only reassign start value if it has been overwritten". I really don't get it. @mahge do you?

In any case, first running `data->callback->updateBoundParameters(data, threadData);` and then running `data->callback->updateBoundVariableAttributes(data, threadData);` with a new provision that overwrites those bound values with start values doesn't seem a good idea to me. On the other hand, @mahge tried to reverse the order they are called and still got problems, so we are probably missing something.

I would suggest to ask @AnHeuermann and @phannebohm if they have some advice about how to avoid overwriting the bound parameter values, while not breaking other stuff.

**Comment by casella** [comment]
BTW, I'm not sure if the point of "Only reassign start value if it has been overwritten" was to fix some mechanism that led to wrong results, or just to spare some time during initialization. In the latter case, maybe the best thing we could do would be to simply revert this commit and see what happens.

**Comment by mahge** [comment]
> Before https://github.com/OpenModelica/OpenModelica/commit/4f66de934b7d6ecfbde869651d3ee7853dd34c13, setAllVarsToStart(data); was finally called again, I'm not sure why. Probably because some start values may have been overwritten, but how? I have no idea. After that commit, this call was removed, but code was added to the template so that every time a bound variable (which apparently also includes bound parameters such as drive.limTau.wMax) is updated, the variable is initialized to its start value. Given the title of https://github.com/OpenModelica/OpenModelica/commit/4f66de934b7d6ecfbde869651d3ee7853dd34c13, the idea was to "Only reassign start value if it has been overwritten". I really don't get it. @mahge do you?

I am taking a blind guess but here is what I think happened. In the beginning we were setting all variables to their start value using `setAllVarsToStart(data)`. Then 4f66de934b7d6ecfbde869651d3ee7853dd34c13 decided we should do that only when we have updated the min value for the variable (I guess: set the variable to its start value only if the variable has a min value set). I do not understand why this is done though. Maybe it makes sense to you. The way it works is by relying on the fact that SimCode has a list `minValueEquations` that it extracted after traversing all variables. For every variable in that list, Codegen will also generated the code that sets it to its start value (after updating the min value).  

>  (which apparently also includes bound parameters such as drive.limTau.wMax) 

This was not the case until #8030. Up until that point we did not add parameters to the list `minValueEquations`. It contained only variables. That is, by the time 4f66de934b7d6ecfbde869651d3ee7853dd34c13 was added only variables had min value assignments generated for them. Now we do it for both. 

> In any case, first running data->callback->updateBoundParameters(data, threadData); and then running data->callback->updateBoundVariableAttributes(data, threadData); with a new provision that overwrites those bound values with start values doesn't seem a good idea to me. On the other hand, @mahge tried to reverse the order they are called and still got problems, so we are probably missing something.

This also probably an unexpected side effect of #8030. Because up until that point there were no parameters in the list of min value equations we maintain.  So, the two functions were completely exclusive. Now we have parameter there. For those parameters `updateBoundVariableAttributes` is overwriting the values. 


There are a few ways to fix this. 
- The first and quickest, is to disable traversal of parameters for extracting min value equations (for non-DO uses). This will not cause any regressions, but it is not preferred because it has its benefits as mentioned in the comments above.  
- Another option is to skip assigning them to their start value if they are parameters. This will not work as far as I can see because the generated code (at least right now) will update the parameter value to its min value first. If we also skip assigning it to the min value as well, then there is no reason to have them in the list in the first place because nothing will be done with them.
- A third option is to keep the parameters and variables in different lists. This is the most logical thing to do except I am not sure what to do with the new list. What exactly are the things we want to do with the collected parameter min. max, nominal, values? I guess for min and max we want to make sure they are within the bounds.

**Comment by casella** [comment]
@mahge so far I had misunderstood how this thing works, it's a bit tricky 😅. Now maybe I get it.

When I see this statement
```C
  data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.min =
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */) ;
```
followed by this one
```
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = 
   data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.start;
```
the way I interpret the first statement is that `data->simulationInfo->realParameter[35]` was used as a buffer to store the result of some operations to infer the min attribute; you then copy that result value to the min attribute, and finally, since the parameter value was used as a buffer, you have to restore it to the original start value. Apparently this works well for variables but not for parameters, for which code is generated to set those values to the binding equation results, which are then lost and replaced by the start attribute.

This way of using data structures seems to me a bit over the edge, because understanding side effects is really difficult (and that's why we are here now), but that's the only way I can interpret it. I guess it was done this way so that one could reuse the same code to infer min, max, and nominal, because for all three you actually work on the variable value itself, otherwise that code would have had to be written three times, once using var.min, once using var.max, and once using var.nominal.

Do you agree?

**Comment by casella** [comment]
@mahge, I think I got it right in the previous comment.

In M_08_bnd.c I see:
```C
static void M_eqFunction_101(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  const int equationIndexes[2] = {1,101};
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = DIVISION((data->simulationInfo->realParameter[33]/* drive.limTau.powMax PARAM */) ,(data->simulationInfo->realParameter[34]/* drive.limTau.tauMax PARAM */) ,"drive.limTau.tauMax");
  TRACE_POP
}
OMC_DISABLE_OPT
int M_updateBoundVariableAttributes(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  /* min ******************************************************** */
  M_eqFunction_101(data, threadData);
  
  infoStreamPrint(LOG_INIT, 1, "updating min-values");
  data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.min = (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */) ;
    infoStreamPrint(LOG_INIT_V, 0, "%s(min=%g)", data->modelData->realParameterData[35].info /* drive.limTau.wMax */.name, (modelica_real) data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.min);
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.start;
  if (ACTIVE_STREAM(LOG_INIT)) messageClose(LOG_INIT);
```
so, `updateBoundVariableAttributes()` computes the min attribute corresponding to this line of code of LimTau:
```modelica
   parameter Modelica.Units.SI.AngularVelocity wMax(min = powMax/tauMax) = 1500 "Maximum speed";
```
in `M_eqFunction_101()`, putting it into `data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */`, then copies it from there into the .min attribute of `drive.limTau.wMax`.

I guess the reason why #8030 did not generate regressions in the library testsuite is that people rarely put min/max attributes with nontrivial expressions in models. However, that is perfectly legitimate, and it would be a shame not to enforce them.

**Comment by casella** [comment]
GOTCHA! @mahge I think I know how to fix this issue. There are two places in initialization.c with these two lines:
```C
    data->callback->updateBoundParameters(data, threadData);
    data->callback->updateBoundVariableAttributes(data, threadData);
```
As we said, this doesn't work, because `updateBoundVariableAttributes` uses the memory cells of the parameter values as a buffer to compute equation-bound min/max/nominal attributes, and then restores them to the start attribute, thus losing the values previously written by `updateBoundParameters`.

If we change it to
```C
    data->callback->updateBoundVariableAttributes(data, threadData);
    data->callback->updateBoundParameters(data, threadData);
```
this issue should then be resolved. However, this doesn't work.

The reason is that updateBoundVariableAttributes needs to have the bound parameters computed in the first place, in order to be able to compute the min/max/nominal equation-bound values. In this case, it needs `drive.limTau.powMax` and `drive.limTau.tauMax` to have the right values, which require to run `updateBoundParameters` first.

So, a simple way to fix this is just to call `updateBoundParameters` first, and then call it again after `updateBoundVariableAttributes` to restore the values accidentally reset to start by the min/max/nominal equations 😎 
```C
    data->callback->updateBoundParameters(data, threadData);
    data->callback->updateBoundVariableAttributes(data, threadData);
    data->callback->updateBoundParameters(data, threadData);
    /* The first call sets the values of bound parameters.
       The second call computes min/max/nominal attributes of variables and parameters (after #8030)
       bound by parameter expressions. The functions that compute those attributes use the variable or parameter
       cell memory to initially store the result, which is then copied to the attribute afterwards. Then, they
       restore the variable or parameter value to its start attribute.
       The third call is made to reset those bound parameters whose values were incorrectly reset 
       to their start attribute */
```
I understand this is not particularly efficient, because in fact we only need to rewrite those parameters that have min/max/nominal equations. However, I guess `updateBoundParameters` is very fast to execute anyway, so I won't bother too much. If there is an easy way to determine if the list of min/max/nominal equations is empty, then we could use it to skip the third call when it is clearly unnecessary.

Can you please try this out?

I would also add those comments, just in case, because figuring this out is not trivial 😅

**Comment by casella** [comment]
Now that I think of it, there could be a much better and more efficient alternative. 

When handling the min/max/nominal equations, the template generates functions such as 

```C
static void M_eqFunction_101(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  const int equationIndexes[2] = {1,101};
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = DIVISION((data->simulationInfo->realParameter[33]/* drive.limTau.powMax PARAM */) ,(data->simulationInfo->realParameter[34]/* drive.limTau.tauMax PARAM */) ,"drive.limTau.tauMax");
  TRACE_POP
}
```
_Only in case the variable is actually a parameter_, when this function is called, the parameter value will already have been set by the functions called by updateBoundParameters. We could then store it in the start attribute, so that updateBoundVariableAttributes restores it properly afterwards, e.g.:
```C
static void M_eqFunction_101(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  const int equationIndexes[2] = {1,101};
  (data->modelData->realParameterData[35].attribute /* drive.limTau.wMax */.start) = data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */;  
  (data->simulationInfo->realParameter[35]/* drive.limTau.wMax PARAM */)  = DIVISION((data->simulationInfo->realParameter[33]/* drive.limTau.powMax PARAM */) ,(data->simulationInfo->realParameter[34]/* drive.limTau.tauMax PARAM */) ,"drive.limTau.tauMax");
  TRACE_POP
}
```
No need to change initialization.c in this case, only a couple of changes to CodegenC.tpl. Note that in case of bound parameters, their start attribute is totally irrelevant, so once we are OK using the parameter value as a buffer to compute min/max/nominal attributes, I don't see anything wrong in using the start attribute as a buffer for the actual bound value 😅 .

I would just put a comment in the .tpl file explaining why we do this, just in case.

Maybe I would try this solution first. I would hear from @AnHeuermann, @kabdelhak and @phannebohm before committing. And, anyway, I would not rush this into 1.20.0, rather into 1.20.1 once we've tested it for some time in the nightly.

**Comment by mahge** [comment]
Sorry for disappearing but it seems like you are having fun debugging this (maybe a bit too much🙂)

> the way I interpret the first statement is that `data->simulationInfo->realParameter[35]` was used as a buffer to store the result of the computation of the min attribute; you then copy that result value to the min attribute, and finally, since the parameter value was used as a buffer, you have to restore it to the original start value. Apparently this works well for variables but not for parameters, for which code is generated to set those values to the binding equation results, which are then lost and replaced by the start attribute.

I too think this is the intention.  

> Apparently this works well for variables but not for parameters, for which code is generated to set those values to the binding equation results, which are then lost and replaced by the start attribute.

Maybe this is what needs to be done for initialization stage. The variables will be updated by their corresponding casualized equations later anyway. 


> This way of using data structures seems to me a bit over the edge, because understanding side effects is really difficult (and that's why we are here now), but that's the only way I can interpret it. I guess it was done this way so that one could reuse the same code to infer min, max, and nominal, because for all three you actually work on the variable value itself, otherwise that code would have had to be written three times, once using var.min, once using var.max, and once using var.nominal.

I think it is just that it is done by different people at different times. Plus, I expect the way Susan Templates work had to have contributed to this style.


> No need to change initialization.c in this case, only a couple of changes to CodegenC.tpl. Note that in case of bound parameters, their start attribute is totally irrelevant, so once we are OK using the parameter value as a buffer to compute min/max/nominal attributes, I don't see anything wrong in using the start attribute as a buffer for the actual bound value 😅 .

> I would just put a comment in the .tpl file explaining why we do this, just in case.

I agree with everything that lead to this and your analysis. The suggested fixes should all work fine. Some might be better for clarity and performance but they all should all work nonetheless.

The question is what do we actually do with the computed min, max ... values for parameters right now? I don't think they are used anywhere. 
##### “Your scientists were so preoccupied with whether they could, they didn't stop to think if they should.” 🙂

> Maybe I would try this solution first. I would hear from @AnHeuermann, @kabdelhak and @phannebohm before committing. And, anyway, I would not rush this into 1.20.0, rather into 1.20.1 once we've tested it for some time in the nightly.

For 1.20.0, we can also disable the traversal for parameters and afterwards do a proper fix and actually use the values we compute for something.

**Comment by casella** [comment]
> I agree with everything that lead to this and your analysis. The suggested fixes should all work fine. Some might be better for clarity and performance but they all should all work nonetheless.

OK. I think the template solution is preferrable, but please go ahead with what you think is the best solution.

> The question is what do we actually do with the computed min, max ... values for parameters right now? I don't think they are used anywhere.
> ##### “Your scientists were so preoccupied with whether they could, they didn't stop to think if they should.” 🙂

Ops. Aren't we checking them at initialization?

> For 1.20.0, we can also disable the traversal for parameters and afterwards do a proper fix and actually use the values we compute for something.

But this would break again the optimization applications that were the subject #7969, right?

Anyway, I think we should release 1.20.0 now, otherwise it becomes a joke. There will always be a new fix that has to be in there... We can think of putting this stuff in 1.20.1, with proper testing and no hurry.

**Comment by mahge** [comment]
> But this would break again the optimization applications that were the subject https://github.com/OpenModelica/OpenModelica/issues/7969, right?

We can enable it if we detect dynamic optimization flags. Otherwise, it will be disabled. I think it should be possible to do this safely.

> Anyway, I think we should release 1.20.0 now, otherwise it becomes a joke. There will always be a new fix that has to be in there... We can think of putting this stuff in 1.20.1, with proper testing and no hurry.

Okay. I will try to add the optional disabling and test it out if you want. If it works out and we have no regressions, we can port it. Otherwise, we can take the time and do the right thing later.

**Comment by mahge** [comment]
This has been fixed in #9853. For now, we have disabled the extraction of min. max and nominal value assignments from parameters. Once we decide what we exactly we want to do with them I can add back the support for it.

I am going to leave the ticket open until we properly address the issue.

**Comment by mahge** [comment]
> OK. This does not hold for min/max attributes of inputs, so that the fix in https://github.com/OpenModelica/OpenModelica/pull/8030 should still work, right?

It will be enabled for cases like #8030 because it checks if we are solving a dynamic optimization problem. Right now, it checks for flags `--optimica` and `-d=gDynOpt` and will be enabled if either of them are set.

> Let's see how it goes with the library testsuite. If this is OK, then we can cherry-pick it for [1.20.1](https://github.com/OpenModelica/OpenModelica/milestone/43).

We already have it [here](https://libraries.openmodelica.org/branches/history/master/2022-12-04%2022:25:55..2022-12-05%2019:12:20.html). There are no relevant regressions.

> OK. My preference goes to changing the template so that the values of bound parameters are backed up in the start attribute, and then later retrieved from there. But I agree we should keep this development for 1.21.0, not for the bugfix release.

It is not really about how to get the values correctly 🙂. It is what to do with them once we achieve that. If we are not doing for them what we are doing for variables (because we saw and agreed that is not correct), then there is nothing being done with the extracted information. We just compute and save these values, but they are never used as far as I can tell.

**Comment by casella** [comment]
OK, then I guess we should also implement checking parameters against their min/max attributes, and issue a runtime warning if the are outside those bounds.

@mahge would you mind creating a simple MWE and opeining a separate ticket for that? Thanks!

**Comment by casella** [comment]
@mahge I'm not sure what you mean, though. If I run this MWE:
```
model TestMaxParameter
  model Foo
    parameter Real p;
    parameter Real q;
    parameter Real r(min = p+q);
  end Foo;
  
  Foo foo(p = 10, q = 5, r = 1);
end TestMaxParameter;
```
I get:
```
The following assertion has been violated at time 0.000000
foo.r >= foo.p + foo.q
Variable violating min constraint: foo.p + foo.q <= foo.r, has value: 1
```
So it seems to me that those parameter min attributes are used indeed. Do I miss something?

**Comment by mahge** [comment]
> So it seems to me that those parameter min attributes are used indeed. Do I miss something?

I am guessing you are using 1.20.0 branch (from your assert message format), so you have something that does not include the "disabled" behavior. I tested the model and I get the assert even with the disabled extraction of parameter `min`, `max` values.

Apparently there is another mechanism that extracts, sets and test these values for parameters. What I noticed was that we never used the values we computed with the now-disabled extraction path. I will try to figure out how this works and let you know.

**Comment by mahge** [comment]
Okay, this seems to be doing more or less what we want to do but in a more straightforward manner. It actually creates dedicated assert statements for the parameter min-max values 
```c
/*
equation index: 1
type: ALGORITHM

  assert(foo.r >= foo.p + foo.q, "Variable violating min constraint: foo.p + foo.q <= foo.r, has value: " + String(foo.r, "g"));
*/
OMC_DISABLE_OPT
static void TestMaxParameter_eqFunction_1(DATA *data, threadData_t *threadData)
...
```

and these are called from `updateBoundParameters()`. Essentially it is doing what we said should be done but without trying to actually set the `.min` attribute of the parameter representation in our generated code. Which kind of makes sense to me since those values are never read. I am not sure what else in the simulation process needs to re-read these `.min` values later.  Is there a situation like that (otger than maybe DO purposes)?

I tested a bit more variations and it seems the current mechanism seems robust and correct. It works even for cases like this
```mo
model TestMaxParameter
  model Foo
    parameter Real p;
    parameter Real q;
    parameter Real r(min = p) = q;
  end Foo;

  Foo foo(p = 10, q = 4);
end TestMaxParameter;
```

I think even the DO case needs to be revised to use this approach instead. Maybe setting the `.min`, `.max` values inside these functions ( the ones generated for the asserts) is better than doing yet another traversal. It is re-extracting information that has already been extracted somewhere else.

**Comment by casella** [comment]
> I am guessing you are using 1.20.0 branch (from your assert message format), 

Yes, actually 1.20.0-dev.beta4

> so you have something that does not include the "disabled" behavior. I tested the model and I get the assert even with the disabled extraction of parameter `min`, `max` values.

OMG. Something so intricate should indeed be documented, shouldn't it?
 
> Apparently there is another mechanism that extracts, sets and test these values for parameters. What I noticed was that we never used the values we computed with the now-disabled extraction path. I will try to figure out how this works and let you know.

OK.

**Comment by casella** [comment]
> Okay, this seems to be doing more or less what we want to do but in a more straightforward manner. It actually creates dedicated assert statements for the parameter min-max values
> 
> ```c
> /*
> equation index: 1
> type: ALGORITHM
> 
>   assert(foo.r >= foo.p + foo.q, "Variable violating min constraint: foo.p + foo.q <= foo.r, has value: " + String(foo.r, "g"));
> */
> OMC_DISABLE_OPT
> static void TestMaxParameter_eqFunction_1(DATA *data, threadData_t *threadData)
> ...
> ```
> 
> and these are called from `updateBoundParameters()`. Essentially it is doing what we said should be done but in the Backend itself (by adding assert statements) and then doesn't bother to actually set the `.min` attribute of the parameter representation in our generated code. 

Sounds reasonable _for bound parameters_, but not for fixed = false parameters, see below. And anyway, I find it a bit creepy that we have data structures (the min/max/nominal record fields in the simulation runtime) that are there but we don't use, and keep on working around that. 

> Which kind of makes sense to me since those values are never read. I am not sure what else in the simulation process needs to re-read these `.min` values later. Is there a situation like that?

Yes, if the parameters have fixed = false and are computed with initial equations, the nominal attribute must be used for scaling if those parameters are iteration variables of nonlinear implicit equations, and the min/max attributes would have to be checked at the end of the initial equations solution process to assess whether it is good or not.

For example, this MWE:
```
model ParameterMinMaxNominal
   parameter Real W_max = 1e9;
   parameter Real part_load = 0.2;
   parameter Real W_nom(nominal = 1e9, min = W_max*part_load, max = W_max, fixed = false);
initial equation
   W_nom + 0.01*Wnom*sin(W_nom/Wmax) = 1e9;
end ParameterMinMaxNominal;
```
should use the W_nom.nominal = 1e9 information to scale the initial nonlinear equation correctly, and should then fail because the solution W_nom = 1.00843e9 is above W_max.
If we are not yet doing that, we definitely should.

In fact, one may also use min/max attributes it for advanced pseudo-Newton codes that take limits on the unknowns into account when setting the damping factor. 

> (other than maybe DO purposes)?

For DO, the parameters could be also unknown, either because they are computed by initial equations, or because they are actually constrained optimization variables.

> I tested a bit more variations and it seems the current mechanism seems robust and correct. 

It certainly is for bound parameters, but I guess not for fixed = false parameters. For them, we need to use the min/max/nominal attribute to set up and check the initial equation process. The fact that we sill don't do that is not a good reason to develop a solution that prevents doing that in the future 😅 
It works even for cases like this

> I think even the DO case needs to be revised to use this approach instead. **Maybe setting the `.min`, `.max` values inside these functions ( the ones generated for the asserts) is better than doing yet another traversal**. It is re-extracting information that has already been extracted somewhere else.

This may be a good idea. The important point is that we have the min/max/nominal attributes set in the data structures, because people writing the runtime may want to use them and will probably assume they are set to the appropriate values.

**Comment by mahge** [comment]
Alright. I am going to try and do this. The first thing step is to make sure these attributes are set properly and are made available without any additional traversal.  If that works well then we can see if the values are used (maybe we were just using garbage values before). If they are not used yet then we call on @AnHeuermann @phannebohm and add the usage in the proper places. If everything goes well then we can update the dynamic optimization code as well.

**Comment by casella** [comment]
@mahge please report the status, this is the [last pending issue](https://github.com/orgs/OpenModelica/projects/9/views/1) before we can release 1.21.0. Thanks!

**Comment by casella** [comment]
@mahge, the buildings libraries didn't like that for some reason, I'm afraid, see [regression report](https://libraries.openmodelica.org/branches/history/master/2023-03-25%2006:03:39..2023-03-26%2011:26:09.html).

**Comment by casella** [comment]
@mahge, the buildings libraries didn't like that for some reason, I'm afraid, see [regression report](https://libraries.openmodelica.org/branches/history/master/2023-03-25%2006:03:39..2023-03-26%2011:26:09.html).

**Comment by mahge** [comment]
I am investigating. I have a suspicion what the issue is. I am trying to verify it. Hopefully it can be back to normal today or tomorrow.

**Comment by mahge** [comment]
I am investigating. I have a suspicion what the issue is. I am trying to verify it. Hopefully it can be back to normal today or tomorrow.

**Comment by casella** [comment]
OK. Technically speaking this would be another ticket, but I guess we can stick to this one. Once you find a fix, please also check that the original issue remains solved. Thanks!

**Comment by mahge** [comment]
We can keep it open instead of opening a new one. If everything goes well it should be fixed now. Github is having some issues so I can not push or open a PR right now. Idk if it is just for me.

**Comment by casella** [comment]
Looks good now, see [regression report](https://libraries.openmodelica.org/branches/history/master/2023-03-26%2011:26:09..2023-03-27%2022:37:37.html).
